core/src/main/java/org/apache/accumulo/core/iterators/system/LocalityGroupIterator.java

```diff
@@ -97,75 +133,117 @@ public static final int seek(HeapIterator hiter, LocalityGroup[] groups, Set<Byt
     else
       cfSet = Collections.emptySet();
 
-    for (LocalityGroup lgr : groups) {
-      // when include is set to true it means this locality groups contains
-      // wanted column families
-      boolean include = false;
+    // determine the set of groups to use
+    Set<LocalityGroup> groupsToUse = new HashSet<LocalityGroup>();
 
-      if (cfSet.size() == 0) {
-        include = !inclusive;
-      } else if (lgr.isDefaultLocalityGroup && lgr.columnFamilies == null) {
-        // do not know what column families are in the default locality group,
-        // only know what column families are not in it
+    // if no column families specified, then include all groups unless !inclusive
+    if (cfSet.size() == 0) {
+      if (!inclusive) {
+        groupsToUse.addAll(Arrays.asList(groups.groups));
```

Comment:
For this case we could have an immutable set that is always used.  Create this once in `LocalityGroupContext`.  Then we can avoid allocating the `HashSet` for `groupsToUse`, just allocate it in the else.  We can also avoid populating the hashset.  Could use Guava immutable set